mlc_llm/relax_model/llama.py

```diff
@@ -556,12 +577,245 @@ def attention_fwd(
+        self.gate = Linear(
+            in_features=config.hidden_size,
+            out_features=config.num_local_experts,
+            bias=False,
+            dtype=config.dtype,
+        )
+        self.num_experts_per_tok = config.num_experts_per_tok
+        self.num_experts = config.num_local_experts
+
+    def topk(self, x, is_ascend, index_dtype, k=-1):
+        # topk along axis -1
+        result = nn.emit(
+            relax.call_dps_packed(
+                "tvm.contrib.thrust.sort_dps",
+                [x, is_ascend],
+                out_sinfo=[
+                    x.struct_info,
+                    relax.TensorStructInfo(x.struct_info.shape, index_dtype),
+                ],
+            )
+        )
+        sorted_x = relax.TupleGetItem(result, 0)
+        indices = relax.TupleGetItem(result, 1)
+        if k != -1:
+            ndim = len(x.struct_info.shape)
+            beg = [0] * ndim
+            end = [x.struct_info.shape[i] for i in range(ndim - 1)] + [k]
+            axes = list(range(ndim))
+            sorted_x = nn.emit(
+                relax.op.strided_slice(sorted_x, axes, beg, end, assume_inbound=True)
+            )
+            indices = nn.emit(relax.op.strided_slice(indices, axes, beg, end, assume_inbound=True))
+        return sorted_x, indices
+
+    def compute_rows_before(self, sorted_expert_ids):
+        return nn.emit(
+            relax.call_dps_packed(
+                "moe_compute_rows_before",
+                [sorted_expert_ids],
+                out_sinfo=relax.TensorStructInfo([self.num_experts], "int64"),
+            )
+        )
+
+    def scatter(self, linear_out, indices):
+        return nn.emit(
+            relax.call_dps_packed(
+                "scatter",
+                [linear_out, indices],
+                out_sinfo=linear_out.struct_info,
+            )
+        )
+
+    def get_token_indices(self, indices):
+        def te_compute(x):
+            return tvm.te.compute(
+                x.shape,
+                lambda *idx: tvm.tir.indexdiv(
+                    x(*idx), tvm.runtime.const(self.num_experts_per_tok, dtype="int32")
+                ).astype("int32"),
+            )
+
+        return nn.emit_te(te_compute, indices)
+
+    def forward(self, hidden_states):
+        hidden_states_shape = hidden_states.struct_info.shape
+        hidden_size = hidden_states_shape[-1]
+        # reshape to 2D
+        hidden_states = nn.emit(relax.op.reshape(hidden_states, (-1, hidden_size)))
+
+        gate = self.gate(hidden_states)
+        scores = nn.emit(relax.op.nn.softmax(gate, axis=-1))
+
+        expert_weights, expert_indices = self.topk(
+            scores, is_ascend=False, k=self.num_experts_per_tok, index_dtype="int32"
+        )  # (num_tokens, top_k), (num_tokens, top_k)
+        expert_weights = nn.emit(expert_weights / R.sum(expert_weights, axis=-1, keepdims=True))
+        flattened_indices = nn.emit(relax.op.flatten(expert_indices))
```

Comment:
This flattening causes the peak VRAM footprint for the intermediate activation to be multiplied by # experts, correct? For large batch size and large # experts, this can be problematic. 

Although that allows one matmul to compute all top-k experts results, I wonder if it can be beneficial to compute it sequentially to reduce VRAM usage and memory traffic. And how much perf drops if we do that.